### ISSUE TITLE
Top createElement / h caller

### ISSUE THREAD

**Issue by luwes** [ungap/babel-plugin-transform-hinted-jsx#2]
This is really cool! JSX should have this feature!

Was wondering is there 1 more piece of information lost with using JSX compared to tagged templates, 
the ability to know the top createElement | h caller (root node) ?

Or is this less of an issue by using a wrapper function?  

Related issue 
https://github.com/developit/htm/issues/143

**Comment by WebReflection** [comment]
> the ability to know the top createElement | h caller (root node) ?

every *outer* template uses a tag, and that tag can be used to relate its template literal to the `createElement` that comes after.

the very top element though, is defined by the render function, but all hints are there.

```jsx
const div = (
  <div>
    <p />
  </div>
);
```

transform into:

```js
const div = React.createElement``('div', null,
  //                           ^^ here: note the tag
  React.createElement('p', null)
  //                 ^ no tag here: static content of the outer el/comp
);
```

That allows *createElement* to behave differently if arguments length is 1 and bind that received template as *createElement* context so that *createElement* there could be smarter if the context is not React or undefined.

In short, I think all hints are already in place.

**Comment by luwes** [comment]
Ah I see, I missed that it was a tag. Thanks for the explanation. Awesome work!

**Comment by WebReflection** [comment]
To clarify, we explored the React.root(…) extra pragma but that both makes migration harder and it doesn’t provide a weak template reference out of the box the engine would know when to collect so while a React.template`` would be great, for testing and PoC of view recycling createElement seemed to be good enough
